Documentation/botr/ryujit-overview.md

```diff
@@ -246,139 +248,140 @@ As the JIT has evolved, changes have been made to improve the ability to reason
+      ▌  STORE_LCL_VAR double V07 cse1         
+t51 =    LCL_VAR   double V07 cse1          $145
+     ┌──▌  t13    double 
+     ├──▌  t51    double 
+t17 = ▌  DIV       double $146
+ t0 =    LCL_VAR   byref  V03 arg3         u:1 (last use) $c0
+     ┌──▌  t0     byref  
+     ├──▌  t17    double 
+      ▌  STOREIND  double
+```
 
 ## <a name="lowering"/>Lowering
 
 Lowering is responsible for transforming the IR in such a way that the control flow, and any register requirements, are fully exposed.
 
-It accomplishes this in two passes.
-
-The first pass is an execution-order traversal that performs context-dependent transformations such as expanding switch statements (using a switch table or a series of conditional branches), constructing addressing modes, etc.  For example, this:
-
-    t0   = lclVar     ref    V00 arg0
-    t1   = lclVar     int    V03 loc1
-           ┌──▌  t1   int
-    t2   = cast       long <- int
-    t3   = const      long   2
-           ┌──▌  t2   long
-           ├──▌  t3   long
-    t4   = <<         long
-           ┌──▌  t0   ref
-           ├──▌  t4   long
-    t5   = +          byref
-    t6   = const      long   16
-           ┌──▌  t5   ref
-           ├──▌  t6   long
-    t7   = +          byref
-           ┌──▌  t7   ref
-    t8   = indir      int
-
-Is transformed into this, in which the addressing mode is explicit:
-
-    t0   = lclVar     ref    V00 arg0
-    t1   = lclVar     int    V03 loc1
-           ┌──▌  t1   int
-    t2   = cast       long <- int
-           ┌──▌  t0   ref
-           ├──▌  t2   long
-    t7   = lea(b+(i*4)+16) byref
-           ┌──▌  t7   ref
-    t8   = indir      int
-
-The next pass annotates the nodes with register requirements. This is done in an execution order traversal in order to ensure that each node's operands are visited prior to the node itself. This pass may also do some transformations that do not require the parent context, such as determining the code generation strategy for block assignments (e.g. `GT_COPYBLK`) which may become helper calls, unrolled loops, or an instruction like `rep stos`.
+It does an execution-order traversal that performs context-dependent transformations such as
+* expanding switch statements (using a switch table or a series of conditional branches)
+* constructing addressing modes (LEA nodes)
+* determining the code generation strategy for block assignments (e.g. `GT_STORE_BLK`) which may become helper calls, unrolled loops, or an instruction like `rep stos`
+* generating machine specific instructions (e.g. generating the `BT` x86/64 instruction)
+* mark "contained" nodes - such a node does not generate any code and relies on its user to include the node's operation in its own codegen (e.g. memory operands, immediate operands)
+* mark "reg optional" nodes - despite the name, such a node may produce a value in a register but its user does not require a register and can consume the value directly from a memory location
```

Comment:
I'm curious about what you mean by "despite the name" - to me, "reg optional" implies that it can have a register or not - either way.

Reply:
The reg optional flag is set on the node that produces the value but ends up being used by the "use" ref position of that node. So it's really "reg optional uses", except that a use isn't a tangible entity in the IR so it's a bit confusing to call it that.

Reply:
Ah, yes I see that now. I'd love to see us support reg-optional defs as well, but that wasn't quite as straightforward as I'd initial thought - shouldn't be rocket science though.

Reply:
I suppose it might be possible to get rid of `SetRegOptional` from `GenTree` and instead put this logic in LSRA's `BuildNode` so that reg optional gets set directly on the ref position. But then the reg optional logic piggybacks on the containment logic so attempting to move it may result in some duplication.